### ISSUE TITLE
Support for requirements files in multiple locations?

### ISSUE THREAD

**Issue by chaoflow** [peterdemin/pip-compile-multi#128]
Thank you very much for this tool! ~~With everything in one directory it does exactly what I expect it to do!~~

@peterdemin Do you see an easy way to support requirements files in multiple locations?

We have one project with multiple packages and this project is contained within another project which has additional packages. We'd like requirements for all packages to be determined in one go.

Currently we have requirements.in files next to the setup.py, which are read from setup.py to set install_requires.
```
super-project/
  code/package1/
    setup.py
    requirements.in -r ../../sub-project/code/package3/requirements.in
  sub-project/
    code/package2/
      setup.py
      requirements.in
      requirements-develop.in
    code/package3/ -r ../package2/requirements.in
      setup.py
      requirements.in
```

I imagine something like:
```
pip-compile-multiple code/*/requirements*.in sub-project/code/*/requirements*.in
```
and have the txt files created next to the corresponding txt file.

**Comment by peterdemin** [comment]
Hi Florian, as I understand all projects have separate requirements and there are no cross-references.
In this case, I'd recommend to run them independently and, maybe, in parallel. There are two reasons for this, since generating lock files may take significant time:

1) Imaging debugging a dependency problem in `package3`, do you want to recompile everything else each time before seeing if the issue is solved?
2) Since all packages have independent requirements, running them in parallel will bring significant gain.

I have projects that need several invocations of `pip-compile-multi` and I implemented that using `tox.ini` or `Makefile`, both options are great for speed and usability.

To summarize: if there is no benefit from a compilation standpoint in adding the loop over directories inside `pip-compile-multi` I prefer to keep it outside. But maybe I'm missing something? What's the motivation behind this request?

**Comment by chaoflow** [comment]
@peterdemin Hi Peter, thank you for the fast response! The packages do depend on each other, I updated the example above and also added an additional requirements-develop.in.

I might be wrong, but I currently do see the need for all requirement files to be taken into account for the resolver to find a common set of versions, even if they would not interdepend. I want to be able to install all packages into one environment with fixed versions of requirements.

I don't see the need to loop over directories in pip-compile-multi, but rather to accept a list of files on the cli in addition to finding files in a requirements directory. For both the corresponding txt-file would be created next to its corresponding in-file.

What do you think?

**Comment by peterdemin** [comment]
I see. Currently, pip-compile-multi relies on the fact that all files are in the same directory. Just because I never thought about this project layout. Here is the part that will need to be changed:

https://github.com/peterdemin/pip-compile-multi/blob/7bd1968c424dd7ce3236885b4b3e4e28523e6915/pipcompilemulti/environment.py#L73-L83

Instead of building `outfile` path from `basedir + name + outext` it would have to use a path to `infile` and change the extension. I think it's fine to make this change since it will preserve current behavior for a single directory case.
Also, CLI function will need to accept positional arguments with glob expressions to in files:

https://github.com/peterdemin/pip-compile-multi/blob/master/pipcompilemulti/cli_v1.py#L38

I don't see how this feature will break other use cases, so I'll accept PR with this functionality.

But before that, would you consider an alternative solution?

If requirements files are interdependent, it means that these are not Python packages (because Python package can't include a file outside of package directory).
So if all packages are always shipped together and their requirements are interdependent, I guess it's fine to put all requirements files in some master directory. All packages would then reference those compiled files like this:

```
super-project/
  requirements/
    package1.in
    package1.txt
    package2.in
    package2.txt
    package2-dev.in -r package2.in
    package2-dev.txt -r package2.txt
  code/package1/
    setup.py
    requirements.txt -r ../../requirements/package1.txt
  sub-project/
    code/package2/
      setup.py
      requirements.txt -r ../../../requirements/package2.txt
      requirements-develop.txt -r ../../../requirements/package2-dev.txt
    code/package3/
      setup.py
      requirements.txt -r ../../../requirements/package3.txt
```

I don't insist on this. If it doesn't fit your project, you are welcome to open a PR for this feature.

**Comment by chaoflow** [comment]
@peterdemin Thank you for the suggestion. I changed the layout now to look like the following:
```
super-project/
  requirements/
    package1.in
    package1.txt
  code/package1/
    setup.py
    requirements.in -> ../../requirements/package1.in  # this is read by setup.py
    requirements.txt -> ../../requirements/package1.txt
  sub-project/
    requirements/
      develop.in
      develop.txt
      package2.in
      package2.txt
      package3.in
      package3.txt
    code/package2/
      setup.py
      requirements.in -> ../../requirements/package2.in  # this is read by setup.py
      requirements.txt -> ../../requirements/package3.txt
    code/package3/ -r ../package2/requirements.in
      setup.py
      requirements.in -> ../../requirements/package3.in  # this is read by setup.py
      requirements.txt -> ../../requirements/package3.txt
```
The sub-project is available as a separate repository as well and therfore must not contain symlinks to the super-project. The symlinks from the packages to projects' requirements directories are followed when sdists are created and the files are correctly included.

This changes the problem to process requirements files in two directories in one go. Do you think this is better solved by allowing ``-d`` to be supplied multiple times or rather allowing arbitrary files to be passed as arguments, i.e.

```
pip-compile-multi -d requirements -d sub-project/requirements
```
or
```
pip-compile-multi requirements/* sub-project/requirements/*
```
The globbing in the latter case would be handled by the shell and pip-compile-multi would only need to process a list of file.

To me the second approach seems more flexible and we could keep ``-d`` for backwards compatibility.

I think the second approach would also enable to make all ``sub-project/requirements/*`` files available in the super-project via symlinks. The txt-files in the super-project would be ignored, the in-files followed and the corresponding txt-file created next to it in the sub-project.

What do you think? Do you see other solutions?

**Comment by chaoflow** [comment]
@peterdemin I was under the assumption that pip-compile-multi takes all requirements files into account before resolving. However, this seems not to be the case. The following fails as it first resolved click to 7.0 and then fails.
```
requirements/
  base.in
    click
  develop.in
    -r base.in
    click<7
```
In the case of two files this is trivial to handle manually, by placing the constraint in base.in. However, in the case of many packages (not just three as above), I'd like to keep the constraints at the package level, while still resolving one set of requirements for all packages.

Do you see ways to support this in pip-compile-multi? Either way, should I create a new issue for this comment to have it tracked separately?

**Comment by chaoflow** [comment]
@peterdemin One more issue, I updated https://github.com/peterdemin/pip-compile-multi/issues/128#issuecomment-479082772

**Comment by peterdemin** [comment]
> Do you think this is better solved by allowing -d to be supplied multiple times or rather allowing arbitrary files to be passed as arguments.

Now I don't like glob expressions for input files, because they are conflicting with existing options:
```
-d, --directory TEXT   Directory path with requirements files
-i, --in-ext TEXT      File extension of input files
```
While this alternative solution could work on its own, it would be hard to explain both options in the documentation and will create extra confusion for new users.
On the other hand, allowing `-d` to be passed multiple times looks acceptable.

> I was under the assumption that pip-compile-multi takes all requirements files into account before resolving.

I see two reasons why `pip-compile-multi` doesn't consider reversed references when resolving package versions:
1. That could make packages in `base.txt` to be implicitly locked to some lower version than expected.
I believe it's better to have constraints explicit in `base.in` file. I also usually add a comment that describes the condition to remove the constraint in the future. 
2. That solution would be much more complex and error-prone.

**Comment by peterdemin** [comment]
Added support for this use-case given that projects cross-reference each other in #221
